subcommands/docker/cmd.go

```diff
@@ -0,0 +1,117 @@
+
+	"github.com/foundriesio/fioctl/subcommands"
+	"github.com/mitchellh/go-homedir"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+const DOCKER_CREDS_HELPER = "docker-credential-fio"
+
+var (
+	dockerConfigFile string
+	helperPath       string
+)
+
+func NewCommand() *cobra.Command {
+	path, err := exec.LookPath("docker")
+	subcommands.DieNotNil(err, "Docker not found on system")
+	helperPath = filepath.Dir(path)
+
+	dockerConfigFile = dockerConfigPath()
+
+	cmd := &cobra.Command{
+		Use:   "configure-docker",
+		Short: "Configure a hub.foundries.io Docker credential helper",
+		Run:   doDockerCreds,
+	}
+	cmd.Flags().StringVarP(&helperPath, "creds-path", "", helperPath, "Path to install credential helper")
+	cmd.Flags().StringVarP(&dockerConfigFile, "docker-config", "", dockerConfigFile, "Docker config file to update")
+	return cmd
+}
+
+func findSelf() string {
+	self := os.Args[0]
+	if !filepath.IsAbs(self) {
+		logrus.Debugf("Looking up path to %s", self)
+		var err error
+		self, err = exec.LookPath(self)
+		subcommands.DieNotNil(err)
+		self, err = filepath.Abs(self)
+		subcommands.DieNotNil(err)
+	}
+	return filepath.Clean(self)
+}
+
+func dockerConfigPath() string {
+	sudoer := os.Getenv("SUDO_USER")
+	if len(sudoer) > 0 {
+		u, err := user.Lookup(sudoer)
+		subcommands.DieNotNil(err)
+		return filepath.Join(u.HomeDir, ".docker/config.json")
+	}
+	path, err := homedir.Expand("~/.docker/config.json")
+	subcommands.DieNotNil(err)
+	return path
+}
+
+func doDockerCreds(cmd *cobra.Command, args []string) {
+	self := findSelf()
+
+	var config map[string]interface{}
+	bytes, err := ioutil.ReadFile(dockerConfigFile)
+	if os.IsNotExist(err) {
+		dockerConfig := filepath.Dir(dockerConfigFile)
+		if _, err := os.Stat(filepath.Dir(dockerConfig)); os.IsNotExist(err) {
+			fmt.Println("Creating docker config directory:", dockerConfig)
+			subcommands.DieNotNil(os.Mkdir(dockerConfig, 0o755))
+		}
+		config = make(map[string]interface{})
+	} else {
+		subcommands.DieNotNil(json.Unmarshal(bytes, &config))
+	}
+
+	helpers, ok := config["credHelpers"]
+	if !ok {
+		config["credHelpers"] = map[string]string{
+			"hub.foundries.io": "fio",
+		}
+	} else {
+		helpers.(map[string]interface{})["hub.foundries.io"] = "fio"
+	}
+
+	configBytes, err := json.MarshalIndent(config, "", "  ")
+	subcommands.DieNotNil(err)
+
+	dst := filepath.Join(helperPath, DOCKER_CREDS_HELPER)
+	fmt.Println("Symlinking", self, "to", dst)
+	subcommands.DieNotNil(os.Symlink(self, dst))
+
+	fmt.Println("Adding hub.foundries.io helper to", dockerConfigFile)
+	subcommands.DieNotNil(ioutil.WriteFile(dockerConfigFile, configBytes, 0o600))
+}
+
+func RunCredsHelper() int {
+	subcommands.Login(NewCommand())
+	creds := struct {
+		Username string
+		Secret   string
+	}{
+		Username: "_fioctl_dch",
+		Secret:   subcommands.Config.ClientCredentials.AccessToken,
+	}
+
+	bytes, err := json.Marshal(creds)
+	subcommands.DieNotNil(err)
+	os.Stdout.Write(bytes)
+	return 0
```

Comment:
The https://docs.docker.com/engine/reference/commandline/login/#credential-helpers documentation is very sparse, but what I understood from it as well as the google diagrams (see my other comments) - docker login will send a `docker-credential-fio get` command when it needs to login. Maybe, it makes sense to check for the `get` subcommand here explicitly; and bail out if docker sends some other (unexpected) subcommand.